### ISSUE TITLE
debug assert from wxWidgets when I drag a mpBitmapLayer

### ISSUE THREAD

**Issue by asmwarrior** [GitHubLionel/wxMathPlot#8]
Hi, I have a png image which has size 800*800.

It shows correctly. But if I pan the plot window, some part of the image will move out of the current view, than I got a wx debug assert message box said that:

~~~~
	// Is there any visible region?
	if (d_width > 0 && d_height > 0)
	{
		// Build the scaled bitmap from the image, only if it has changed:
		if (m_scaledBitmap.GetWidth() != d_width || m_scaledBitmap.GetHeight() != d_height || m_scaledBitmap_offset_x != offset_x
				|| m_scaledBitmap_offset_y != offset_y)
		{
			wxRect r(wxRect(offset_x, offset_y, b_width, b_height));
			// Just for the case....
			if (r.x < 0)
				r.x = 0;
			if (r.y < 0)
				r.y = 0;
			if (r.width > m_bitmap.GetWidth())
				r.width = m_bitmap.GetWidth();
			if (r.height > m_bitmap.GetHeight())
				r.height = m_bitmap.GetHeight();

			m_scaledBitmap = wxBitmap(wxBitmap(m_bitmap).GetSubBitmap(r).ConvertToImage().Scale(d_width, d_height));
~~~~

I just checked the code, that the `r` value is

~~~~
[debug]> p r
[debug]$2 = {x = 0, y = 1, width = 800, height = 800}
[debug]>>>>>>cb_gdb:

$2 = {x = 0, y = 1, width = 800, height = 800}
~~~~

While, the assert is from the `GetSubBitmap` function, it said the `y+height` should smaller than 800. Which means the SubBitmap area is beyond the original bitmap.

Any ideas?

Thanks.

**Comment by asmwarrior** [comment]
~~~~
			if (r.x + r.width > m_bitmap.GetWidth())
				r.width = m_bitmap.GetWidth() - r.x;

			if (r.y + r.height > m_bitmap.GetHeight())
				r.height = m_bitmap.GetHeight() - r.y;
~~~~

I add some sanity check here, so that the `r` should not beyond the original bitmap area.

But that is only a sanity check, I don't looked into the details of the calculation.

**Comment by GitHubLionel** [comment]
This part of code is the same of the original component. If your check correct the problem, I can add it to my code.
So I can replace :
`	
		if (r.width > m_bitmap.GetWidth())
				r.width = m_bitmap.GetWidth();
			if (r.height > m_bitmap.GetHeight())
				r.height = m_bitmap.GetHeight();
`
by
`
			if (r.x + r.width > m_bitmap.GetWidth())
				r.width = m_bitmap.GetWidth() - r.x;
			if (r.y + r.height > m_bitmap.GetHeight())
				r.height = m_bitmap.GetHeight() - r.y;
`

**Comment by asmwarrior** [comment]
> This part of code is the same of the original component. If your check correct the problem, I can add it to my code. So I can replace : ` if (r.width > m_bitmap.GetWidth()) r.width = m_bitmap.GetWidth(); if (r.height > m_bitmap.GetHeight()) r.height = m_bitmap.GetHeight(); ` by `if (r.x + r.width > m_bitmap.GetWidth()) r.width = m_bitmap.GetWidth() - r.x; if (r.y + r.height > m_bitmap.GetHeight()) r.height = m_bitmap.GetHeight() - r.y;`

From my point of view, I think this is the correct fix. Please do this, thanks!

**Comment by GitHubLionel** [comment]
OK, it is done.
I also add "Magnetize" property for the mouse moving.

**Comment by asmwarrior** [comment]
I think the drag of bitmap layer issue is fixed.

But I'm not sure what does the "Magnetize" function used for. If I enable this feature. I see there are always two crosses when I move the mouse. Is that correct? One cross is always follow the mouse cursor, another cross is a big slow, and has a motion lag compared with the first cross.

**Comment by GitHubLionel** [comment]
Magnetize is an alternative to read coordinate of the mouse directly over the axis instead of info coord.
You may have some trouble if you use it with a movable object or continuously update the plot because you continuously refresh the plot. Use it with static plot only.

**Comment by GitHubLionel** [comment]
Ok, I add some tests to minimize the impact of refresh when mouse is moving.

**Comment by asmwarrior** [comment]
> Ok, I add some tests to minimize the impact of refresh when mouse is moving.

I just test the latest code, but I see there are still two crosses when I move the mouse. Sorry, I think I use "static plot".  I use a bitmap layer, is it a static plot?

**Comment by asmwarrior** [comment]
> I think the drag of bitmap layer issue is fixed.
> 

Sorry, it looks like this issue is not fixed.

I still got an assert:

~~~~
> p r

[debug]> p r
[debug]$1 = {x = 800, y = 0, width = 0, height = 800}
[debug]>>>>>>cb_gdb:

$1 = {x = 800, y = 0, width = 0, height = 800}
> p d_width

[debug]> p d_width
[debug]$2 = 1
[debug]>>>>>>cb_gdb:

$2 = 1
> p d_height

[debug]> p d_height
[debug]$3 = 121
[debug]>>>>>>cb_gdb:

$3 = 121
~~~~

In the line:

~~~~
	// Is there any visible region?
	if ((d_width > 0) && (d_height > 0))
	{
		// Build the scaled bitmap from the image, only if it has changed:
		if (m_scaledBitmap.GetWidth() != d_width || m_scaledBitmap.GetHeight() != d_height ||
				m_scaledBitmap_offset_x != offset_x || m_scaledBitmap_offset_y != offset_y)
		{
			wxRect r = wxRect(offset_x, offset_y, b_width, b_height);
			// Just for the case....
			if (r.x < 0)
				r.x = 0;
			if (r.y < 0)
				r.y = 0;

			// Correction : https://github.com/GitHubLionel/wxMathPlot/issues/8
			if (r.x + r.width > m_bitmap.GetWidth())
				r.width = m_bitmap.GetWidth() - r.x;
			if (r.y + r.height > m_bitmap.GetHeight())
				r.height = m_bitmap.GetHeight() - r.y;

			if (r.x + r.width > m_bitmap.GetWidth())
				r.width = m_bitmap.GetWidth() - r.x;

			if (r.y + r.height > m_bitmap.GetHeight())
				r.height = m_bitmap.GetHeight() - r.y;

			m_scaledBitmap = wxBitmap(wxBitmap(m_bitmap).GetSubBitmap(r).ConvertToImage().Scale(d_width, d_height));
			m_scaledBitmap_offset_x = offset_x;
			m_scaledBitmap_offset_y = offset_y;
		}
~~~~

You can see, the `r` is wrong here, so I still got assert when calling `GetSubBitmap(r)`.

I think we still need some extra condition check for the `r` before we call the `GetSubBitmap(r)`.

At least `r` should have width and height larger than `0`.

**Comment by asmwarrior** [comment]
> I don't use at all mpBitmapLayer so it is difficult to help you. Could you make a simple projet that reproduce the problem ?

To demonstrate the crash issue, you can add those lines in the `MathPlotConfigDemo.cpp` file:

~~~~
	wxImage bmp = wxImage("test.png", wxBITMAP_TYPE_PNG);
	mpBitmapLayer* bitmapLayer = new mpBitmapLayer();
	bitmapLayer->SetBitmap(bmp, 0, 0, 320, 240);
	m_plot->AddLayer(bitmapLayer);
~~~~

And here is the screencast of the crash in a mp4 file.




https://user-images.githubusercontent.com/561818/235814315-cc1a27f3-e0ee-4291-a18c-27709ec1d479.mp4


I hope you can reproduce this crash.

**Comment by asmwarrior** [comment]
About the two crosses issue, I can have screen cast here, note that I have some custom layer which is derived from the basic wxMathplot layer(such as mpBitmapLayer).


https://user-images.githubusercontent.com/561818/235815451-123ee04e-f6be-4a75-b460-930bb04a2356.mp4

It looks like the two crosses issue does not happen in my previous bitmap crash demo.

**Comment by GitHubLionel** [comment]
Ok, I see that the problem is when the bitmap disappear exactly totally from plot (if you pan fast, you may skip the problem).
So, we just need to test before if size become null.
For the crosses issue, I could not reproduce your video. 
I notice that bitmap is always on top, so he hide the axis. Maybe it will be better if we have the opposite ?

**Comment by asmwarrior** [comment]
> Ok, I see that the problem is when the bitmap disappear exactly totally from plot (if you pan fast, you may skip the problem). So, we just need to test before if size become null. 

OK, I see your fix in the github head, thanks.

> For the crosses issue, I could not reproduce your video. I notice that bitmap is always on top, so he hide the axis. Maybe it will be better if we have the opposite ?

I think I have found the reason of the "two crosses". In my own code, I have a timer to refersh the plot continuously. So that I guess you are using the XOR mode of drawing the Magnetize lines get wrong.

**Comment by GitHubLionel** [comment]
Yes I use XOR for best performance (no need to repaint whole the plot) but indeed, if you refresh between, that a problem. 
But logically, if you use the UpdateAll() method to refresh your plot, it should be avoid the problem.

**Comment by asmwarrior** [comment]
> Yes I use XOR for best performance (no need to repaint whole the plot) but indeed, if you refresh between, that a problem. But logically, if you use the UpdateAll() method to refresh your plot, it should be avoid the problem.

Thanks for the help.

I just change the `Refresh` to `UpdateAll` in the timer event handler, but it looks like I still see the "two crosses" issue.

**Comment by asmwarrior** [comment]
Well, I just looked at this issue. It looks like UpdateAll function just call the Refersh function internally, so I believe they are the same issue here.

I guess the logic here is by using the XOR drawing method of the corss lines, here is the event sequence:

~~~~
MouseMove MouseMove MouseMove Timer MouseMove MouseMove

drawXOR   drawXOR   drawXOR   New   drawXOR   drawXOR
~~~~

So, how do we handle the Timer event? It looks like `drawXOR` is done in the client dc, while when I got a Timer event, I will call the `UpdateAll`, which finally call the draw function in paint dc.

Things are complex.
